### Homework 1
#### Submitted by: Falak Jain

In [11]:
# Importing libraries
import pandas as pd
import re
import warnings
warnings.filterwarnings("ignore")

In [12]:
# Reading dataset
mcd_df = pd.read_csv('../datasets/mcdonalds-yelp-negative-reviews.csv',encoding = 'latin8')

In [13]:
mcd_df.head()

_unit_id     city                                             review
0  679455653  Atlanta  I'm not a huge mcds lover, but I've been to be...
1  679455654  Atlanta  Terrible customer service. I came in at 9:30pm...
2  679455655  Atlanta  First they "lost" my order, actually they gave...
3  679455656  Atlanta  I see I'm not the only one giving 1 star. Only...
4  679455657  Atlanta  Well, it's McDonald's, so you know what the fo...

### 1. 4 recurring themes/topics that reviewers are unhappy with

1. Customers are constantly complaining about the slow service at McDonalds
2. McDonalds often gets customers' orders wrong
3. Patrons are not satisfied with the quality of customer service
4. Customers often complain about the cleanliness and hygiene of the store and food

### 2. Grouping all occurences of phrases and replacing with place holder

In [14]:
# 1. Slow Service Complains
mcd_df["results"] = mcd_df["review"].str.replace(r'\b[sS]low','_SLOW_',case = False)

In [15]:
# 2. Wrong orders
mcd_df["results"] = mcd_df["results"].str.replace(r'\b[Ww]rong','_WRONG_',case = False)
mcd_df["results"] = mcd_df["results"].str.replace(r'\b[Ii]ncorrect','_WRONG_',case = False)

In [16]:
# 3. Quality of customer service
mcd_df["results"] = mcd_df["results"].str.replace(r'\b[Cc]ustomer [Ss]ervice\b','_CUSTOMER_SERVICE_',case = False)

In [17]:
# 4. Cleanliness and hygiene of the store
mcd_df["results"] = mcd_df["results"].str.replace(r'\b[Dd]irt','_DIRTY_',case = False)
mcd_df["results"] = mcd_df["results"].str.replace(r'\b[Ff]ilth','_DIRTY_',case = False)

### 3. Performing a word count at total and city levels

In [18]:
# Creating a flag
mcd_df['SLOW_SERVICE'] = mcd_df['results'].apply(lambda x: 1 if '_SLOW_' in x else 0)
mcd_df['WRONG_ORDER'] = mcd_df['results'].apply(lambda x: 1 if '_WRONG_' in x else 0)
mcd_df['CUSTOMER_SERVICE'] = mcd_df['results'].apply(lambda x: 1 if '_CUSTOMER_SERVICE_' in x else 0)
mcd_df['DIRTY'] = mcd_df['results'].apply(lambda x: 1 if '_DIRTY_' in x else 0)

In [19]:
# Finding count of keywords at a city and aggregate level
city_count_df = mcd_df.groupby('city').sum().drop('_unit_id',axis = 1)
city_count_df.loc['Total'] = city_count_df.sum()
city_count_df

SLOW_SERVICE  WRONG_ORDER  CUSTOMER_SERVICE  DIRTY
city                                                           
Atlanta                23           16                10     12
Chicago                16           23                15      9
Cleveland              16           11                 6      6
Dallas                  7            9                 8      3
Houston                15           10                 8      7
Las Vegas              41           52                25     23
Los Angeles            17           24                 8      8
New York                9            4                11     11
Portland                3            9                 4      5
Total                 147          158                95     84

In [20]:
# Finding proportion of keywords at a city and aggregate level
city_df = mcd_df.groupby('city').mean().drop('_unit_id',axis = 1)
city_df.loc['Mean'] = city_df.mean()
city_df

SLOW_SERVICE  WRONG_ORDER  CUSTOMER_SERVICE     DIRTY
city                                                              
Atlanta          0.176923     0.123077          0.076923  0.092308
Chicago          0.073059     0.105023          0.068493  0.041096
Cleveland        0.225352     0.154930          0.084507  0.084507
Dallas           0.093333     0.120000          0.106667  0.040000
Houston          0.142857     0.095238          0.076190  0.066667
Las Vegas        0.100244     0.127139          0.061125  0.056235
Los Angeles      0.101796     0.143713          0.047904  0.047904
New York         0.054545     0.024242          0.066667  0.066667
Portland         0.030928     0.092784          0.041237  0.051546
Mean             0.111004     0.109572          0.069968  0.060770

### 4. Provide a few sentences with your findings and business recommendations. Make any assumptions you'd like. I just want you to get into the habit of "finishing" your analysis: to avoid delivering technical numbers to a non-technical manager.

Some considerations in your analysis:

Explain what some of pitfalls/limitations are of using only a word count analysis to make these inferences. What additional research/steps would you need to do to verify your conclusions?

1. Customers complained about the slow service in about 11% of the reviews on a national level with Atlanta, Cleveland and Houston having a higher percentage than the national average. I recommend increasing the staff at McDonalds outlets to improve speed and to invest in the training of staff to increase productivity.


2. About 10% of the reviews pertained to the outlet getting a customer's order wrong. I recommend implementing quality check mechanisms and more effective information technology systems in place to prevent incorrect orders.


3. About 7% of reviews included comments on the customer service. Managers should be held accountable towards their outlet's customer service and employees should be trained in speaking in a welcoming and non-confrontational manner with customers to improve customer service standards. Blind tests can be conducted by the corporate office of McDonalds to find areas of improvement.


4. About 6% of reviews were concerned with the cleanliness of outlets. McDonalds should employ better standards to cleanliness, equip their outlets with state of the art cleaning equipment and run regular surprise cleanliness checks at different locations.


5. McDonalds outlets in Cleveland seem to be consistently worse than the average in all of the 4 themes identified. Therefore, McDonalds shall conduct a thorough audit of Cleveland locations and determine effective improvement measures to improve services which would in turn improve reviews.


6. Word count analysis should not be solely relied upon since a review can have the target word in it but the review could be referring to other graver concerns rather than the target phrase that we are looking at. It is also important to consider the sentiment of the review specifically regarding the target phrase. e.g. the word 'customer service' can occur 'n' number of times but we also need to see whether the reviewer is speaking about the word in positive and negative ways.